fmf/utils.py

```diff
@@ -808,51 +783,27 @@ def run(command, cwd=None, check_exit_code=True, env=None):
 #  Convert dict to yaml
 # ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-# Special hack to store multiline text with the '|' style
-# See https://stackoverflow.com/questions/45004464/
-# Python 2 version
-try:  # pragma: no cover
-    yaml.SafeDumper.orig_represent_unicode = yaml.SafeDumper.represent_unicode
-
-    def repr_unicode(dumper, data):
-        if '\n' in data:
-            return dumper.represent_scalar(
-                u'tag:yaml.org,2002:str', data, style='|')
-        return dumper.orig_represent_unicode(data)
-    yaml.add_representer(unicode, repr_unicode, Dumper=yaml.SafeDumper)
-
-# Python 3 version
-except AttributeError:
-    yaml.SafeDumper.orig_represent_str = yaml.SafeDumper.represent_str
-
-    def repr_str(dumper, data):
-        if '\n' in data:
-            return dumper.represent_scalar(
-                u'tag:yaml.org,2002:str', data, style='|')
-        return dumper.orig_represent_str(data)
-    yaml.add_representer(str, repr_str, Dumper=yaml.SafeDumper)
-
-
 def dict_to_yaml(data, width=None, sort=False):
     """ Convert dictionary into yaml """
     output = StringIO()
-    try:
-        yaml.safe_dump(
-            data, output, sort_keys=sort,
-            encoding='utf-8', allow_unicode=True,
-            width=width, indent=4, default_flow_style=False)
-    except TypeError:  # pragma: no cover
-        # FIXME: Temporary workaround for rhel-8 to disable key sorting
-        # https://stackoverflow.com/questions/31605131/
-        # https://github.com/psss/tmt/issues/207
-        def representer(self, data): return self.represent_mapping(
-            'tag:yaml.org,2002:map', data.items())
-        yaml.add_representer(dict, representer, Dumper=yaml.SafeDumper)
-        yaml.safe_dump(
-            data, output, encoding='utf-8', allow_unicode=True,
-            width=width, indent=4, default_flow_style=False)
-    # For Python 2 we need to decode the string
-    try:
-        return output.getvalue().decode('utf-8')
-    except AttributeError:
-        return output.getvalue()
+
+    # Set formatting options
+    yaml = YAML()
+    yaml.indent(mapping=4, sequence=4, offset=2)
+    yaml.default_flow_style = False
+    yaml.allow_unicode = True
+    yaml.encoding = 'utf-8'
+    yaml.width = width
+
+    # Convert multiline strings
+    scalarstring.walk_tree(data)
```

Comment:
@FrNecas, I've copied this from your `tmt` migration to `ruamel` but I'm not sure: Why exactly is this needed? Does it modify `data`? Should we create a deep copy to prevent any changes to `data`? I tried to apply it on a `dict` which contained a multiline string but could not see any change...

Reply:
@psss I guess I should have been more explicit in the comment. Yes indeed, it should modify the `data` as far as I remember. There was a hack in tmt that took care of this for PyYAML (so that multiline strings would be formatted accordingly) so I wanted to keep the behaviour the same and found this: https://stackoverflow.com/a/50853272 .

According to the comment and docs, the `walk_tree` method should convert string to `LiteralScalarString` if the element is a multiline string which is what we wanted due to formatting. I don't know what you've tried but a quick test shows this:

```
>>> from ruamel.yaml import scalarstring
>>> data = {'a': 'a\nb\nc\n'}
>>> type(data['a'])
<class 'str'>
>>> scalarstring.walk_tree(data)
>>> type(data['a'])
<class 'ruamel.yaml.scalarstring.LiteralScalarString'>
>>> data['a']
'a\nb\nc\n'
```
The `LiteralScalarString` class probably has a set `__str__` formatting but as you can see, it was transformed. As for your last question, I don't really know. It may make sense to create a copy in fmf so that the formatting is separate from the inner logic? On the other hand, it inherits from `str` class so it behaves like a normal string just has some more logic related to ruamel so it may not be a problem at all to do this without copying. I will take a closer look tomorrow or on Friday.

Reply:
On the other hand, users of fmf as a library may expect that it keeps the formatting the way it was? It's definitely up for a discussion :)